fdbserver/DataDistribution.actor.cpp

```diff
@@ -3272,6 +3284,143 @@ ACTOR Future<Void> dataDistribution(
+	loop {
+		state ReadYourWritesTransaction tr(cx);
+
+		loop {
+			try {
+				TraceEvent("DataDistributor_MonitorConfigurationStart", self->ddId);
+				tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				tr.setOption( FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE );
+				Standalone<RangeResultRef> results = wait( tr.getRange( configKeys, CLIENT_KNOBS->TOO_MANY ) );
+				ASSERT( !results.more && results.size() < CLIENT_KNOBS->TOO_MANY );
+
+				DatabaseConfiguration conf;
+				conf.fromKeyValues( (VectorRef<KeyValueRef>) results );
+				if ( conf != self->configuration->get() ) {
+					TraceEvent("DataDistributor_UpdateConfiguration", self->ddId).detail("Config", conf.toString());
+					self->configuration->set( conf );
+					self->configurationTrigger.trigger();
+				}
+
+				state Future<Void> watchFuture = tr.watch(configVersionKey);
+				wait( tr.commit() );
+				wait( watchFuture );
+				break;
+			} catch (Error& e) {
+				wait( tr.onError(e) );
+			}
+		}
+	}
+}
+
+static std::set<int> const& normalDataDistributorErrors() {
+	static std::set<int> s;
+	if (s.empty()) {
+		s.insert( error_code_worker_removed );
+		s.insert( error_code_broken_promise );
+		s.insert( error_code_actor_cancelled );
+		s.insert( error_code_please_reboot );
+		s.insert( error_code_movekeys_conflict );
+	}
+	return s;
+}
+
+static std::set<int> const& normalRateKeeperErrors() {
+	static std::set<int> s;
+	if (s.empty()) {
+		s.insert( error_code_worker_removed );
+		s.insert( error_code_broken_promise );
+		s.insert( error_code_actor_cancelled );
+		s.insert( error_code_please_reboot );
+	}
+	return s;
+}
+
+ACTOR Future<Void> dataDistributor(DataDistributorInterface di, Reference<AsyncVar<struct ServerDBInfo>> db ) {
+	state UID lastClusterControllerID(0,0);
+	state Reference<AsyncVar<DatabaseConfiguration>> configuration( new AsyncVar<DatabaseConfiguration>(DatabaseConfiguration()) );
+	state Reference<DataDistributorData> self( new DataDistributorData(db, configuration, di.id()) );
+	state Future<Void> collection = actorCollection( self->addActor.getFuture() );
+
+	TraceEvent("DataDistributor_Starting", di.id());
+	self->addActor.send( waitFailureServer(di.waitFailure.getFuture()) );
+	self->addActor.send( configurationMonitor( self ) );
+
+	loop choose {
+		when ( wait( self->configurationTrigger.onTrigger() ) ) {
+			self->refreshDcIds();
+			break;
+		}
+		when ( wait(self->dbInfo->onChange()) ) {}
+	}
+
+	try {
+		TraceEvent("DataDistributor_Running", di.id());
+		state PromiseStream< std::pair<UID, Optional<StorageServerInterface>> > ddStorageServerChanges;
+		state double lastLimited = 0;
+		state Future<Void> distributor = reportErrorsExcept( dataDistribution( self->dbInfo, di.id(), self->configuration->get(), ddStorageServerChanges, self->primaryDcId, self->remoteDcIds, &lastLimited ), "DataDistribution", di.id(), &normalDataDistributorErrors() );
+		self->addActor.send( distributor );
```

Comment:
do not add distributor to addActor, wait in the choose when below for distributor